### PR TITLE
[feat]: [CDS-79080] Add config files and manifest overrides support for custom stage environment

### PR DESCRIPTION
**Pull request by lovish1999**
## Describe your changes

Added support for manifest and config file overrides for custom stage environment. Moved the common code from custom stage environment and ServiceStepV3 class to a common helper class. Also, added plan creators for manifest and config files steps as children of custom stage environment step. 

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53292)

### CODE REVIEW

**Comment by lovish1999** [comment]
trigger ut0

**Comment by lovish1999** [comment]
trigger ut2

**Comment by lovish1999** [comment]
trigger ut3

**Comment by lovish1999** [comment]
trigger ut4

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentMapper.java` line 111 at r1](https://reviewable.io/reviews/harness/harness-core/53292#-NfGkpUv6wdnRh8EoOTU:-NfGkpUv6wdnRh8EoOTV:b-lz9m7o) ([raw file](https://github.com/harness/harness-core/blob/8b070116e6fa2924b2bfbdaf56671fa495802451/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentMapper.java#L111)):*
> ```Java
>   }
> 
>   public ServiceStepV3Parameters toServiceStepV3Parameters(CustomStageEnvironmentStepParameters parameters) {
> ```

this should probably have a @NotNull annotation of the field. Its very wrong if step parameters are null, we need to be able to catch such cases rather than ignoring them

**Comment by lovish1999** [comment]
trigger uts

**Comment by lovish1999** [comment]
trigger ut5

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java [hunk 1]
```diff
@@ -171,7 +170,7 @@ private Ambiance prepareServiceAmbiance(Ambiance ambiance) {
         return AmbianceUtils.clone(ambiance, i + 1);
       }
     }
-    throw new UnsupportedOperationException("Not inside service step or one of it's children");
+    return ambiance;
```
Comment (tchaurasiya) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java:173]: Why we are removing the exception?  shouldn't we add custom stage in levels

**Review by tchaurasiya — CHANGES_REQUESTED** [review]

**Review by ivan-83 — CHANGES_REQUESTED** [review]
> Reviewed 2 of 9 files at r1, 5 of 9 files at r2, 4 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53292)* status: 12 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/CustomStageEnvironmentStep.java` line 206 at r3](https://reviewable.io/reviews/harness/harness-core/53292#-NfMDEKM5a_dWlrVXee8:-NfMDEKNA_Etoyq5AS6H:b-mrsovz) ([raw file](https://github.com/harness/harness-core/blob/dff9c602c1f081004ed268632d4cfb0c01f4cbd1/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/CustomStageEnvironmentStep.java#L206)):*
> ```Java
>     serviceStepV3Helper.addManifestsOutputToStepOutcome(ambiance, stepOutcomes);
> 
>     serviceStepV3Helper.addConfigFilesOutputToStepOutcome(ambiance, stepOutcomes);
> ```

What about other outputs like app settings and con strings?
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java` line 173 at r3](https://reviewable.io/reviews/harness/harness-core/53292#gh-1338551782:-NfMEmmq1nis8rJDi1uq:bmpu4bv) ([raw file](https://github.com/harness/harness-core/blob/dff9c602c1f081004ed268632d4cfb0c01f4cbd1/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java#L173)):*
<details><summary><i>Previously, tchaurasiya (Tathagat Chaurasiya) wrote…</i></summary><blockquote>

Why we are removing the exception?  shouldn't we add custom stage in levels
</blockquote></details>

Agree. Can we have a different method for the custom stage? Where do we use this method?
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/helper/PlanCreatorHelper.java` line 69 at r3](https://reviewable.io/reviews/harness/harness-core/53292#-NfMFC1S2RXrC6On4BQg:-NfMFC1S2RXrC6On4BQh:bmdyvv0) ([raw file](https://github.com/harness/harness-core/blob/dff9c602c1f081004ed268632d4cfb0c01f4cbd1/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/helper/PlanCreatorHelper.java#L69)):*
> ```Java
>     planCreationResponseMap.put(
>         configFilesNode.getUuid(), PlanCreationResponse.builder().planNode(configFilesNode).build());
>   }
> ```

What about app settings and conn strings? 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java` line 210 at r3](https://reviewable.io/reviews/harness/harness-core/53292#-NfMFSrGArdvtGbcYlK4:-NfMFSrGArdvtGbcYlK5:bk1hfp5) ([raw file](https://github.com/harness/harness-core/blob/dff9c602c1f081004ed268632d4cfb0c01f4cbd1/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java#L210)):*
> ```Java
> 
>     // Add manifests node
>     planCreatorHelper.addManifestsNode(planCreationResponseMap, childrenNodeIds);
> ```

Should we add manifest and config files as children of custom stage env plan creator or here?

**Review by lovish1999 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53292)* status: 12 file(s) to review. 2 change requests. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentMapper.java` line 111 at r1](https://reviewable.io/reviews/harness/harness-core/53292#-NfGkpUv6wdnRh8EoOTU:-NfMa5AW6Kq2c0NZGy49:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/8b070116e6fa2924b2bfbdaf56671fa495802451/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentMapper.java#L111)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

this should probably have a @NotNull annotation of the field. Its very wrong if step parameters are null, we need to be able to catch such cases rather than ignoring them
</blockquote></details>

Done.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/CustomStageEnvironmentStep.java` line 206 at r3](https://reviewable.io/reviews/harness/harness-core/53292#-NfMDEKM5a_dWlrVXee8:-NfMa5eOCr4_EshKSQS0:b-oygo38) ([raw file](https://github.com/harness/harness-core/blob/dff9c602c1f081004ed268632d4cfb0c01f4cbd1/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/CustomStageEnvironmentStep.java#L206)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

What about other outputs like app settings and con strings?
</blockquote></details>

They do not have a step or output. They are just overrides which are consumed by sweeping output at service step and then used later on. 
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java` line 173 at r3](https://reviewable.io/reviews/harness/harness-core/53292#gh-1338551782:-NfMaUEl-d500OU6-Inz:b-wudaka) ([raw file](https://github.com/harness/harness-core/blob/dff9c602c1f081004ed268632d4cfb0c01f4cbd1/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java#L173)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

Agree. Can we have a different method for the custom stage? Where do we use this method?
</blockquote></details>

This method is used in manifest , config files and other childrens of service while getting the logcallback. Yes, as Tathagat suggested, I will add custom stage env. step also in the levels. 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/helper/PlanCreatorHelper.java` line 69 at r3](https://reviewable.io/reviews/harness/harness-core/53292#-NfMFC1S2RXrC6On4BQg:-NfMaiTQ1laahNQ2yuoA:breob04) ([raw file](https://github.com/harness/harness-core/blob/dff9c602c1f081004ed268632d4cfb0c01f4cbd1/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/helper/PlanCreatorHelper.java#L69)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

What about app settings and conn strings?
</blockquote></details>

They do not have a step
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java` line 210 at r3](https://reviewable.io/reviews/harness/harness-core/53292#-NfMFSrGArdvtGbcYlK4:-NfMam9X5pceVN9Dz-nv:b-adri07) ([raw file](https://github.com/harness/harness-core/blob/dff9c602c1f081004ed268632d4cfb0c01f4cbd1/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java#L210)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

Should we add manifest and config files as children of custom stage env plan creator or here?
</blockquote></details>

We are adding them as children of custom stage env. step. It's similar to how its done in service step.

**Review by ivan-83 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53292)* status: 12 file(s) to review. 2 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/CustomStageEnvironmentStep.java` line 206 at r3](https://reviewable.io/reviews/harness/harness-core/53292#-NfMDEKM5a_dWlrVXee8:-NfMmB82-J-VXjvVueAb:bsiomif) ([raw file](https://github.com/harness/harness-core/blob/dff9c602c1f081004ed268632d4cfb0c01f4cbd1/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/CustomStageEnvironmentStep.java#L206)):*
<details><summary><i>Previously, lovish1999 (Lovish Bansal) wrote…</i></summary><blockquote>

They do not have a step or output. They are just overrides which are consumed by sweeping output at service step and then used later on.
</blockquote></details>

Ok, how are we fetching app settings and con strings from git?

**Comment by lovish1999** [comment]
trigger uts

**Review by ivan-83 — COMMENTED** [review]
LGTM with minor comments.

> Reviewed 1 of 1 files at r4, 7 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53292#-:-NfpBYfZ645hpEG2g1r2:b-o9mibo)* status: 13 file(s) to review. 2 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepOverrideHelper.java` line 775 at r5](https://reviewable.io/reviews/harness/harness-core/53292#-Nfp9kce7hYf46fMRhU8:-Nfp9kce7hYf46fMRhU9:b8ngg24) ([raw file](https://github.com/harness/harness-core/blob/04cf7342dd948bad771ae036bb73c2c2e8e85c62/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepOverrideHelper.java#L775)):*
> ```Java
>     }
> 
>     saveFinalAppSettingsToSweepingOutput(ambiance, appSettingsSweepingOutputName, finalAppSettings);
> ```

Minor: we can put saveFinalAppSettingsToSweepingOutput(ambiance, appSettingsSweepingOutputName, finalAppSettings); as part of the above if. We are saving to sweeping output only if the configuration is different than null.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepOverrideHelper.java` line 789 at r5](https://reviewable.io/reviews/harness/harness-core/53292#-NfpAKnU46s3q1HKvSUp:-NfpAKnU46s3q1HKvSUq:bf26u1) ([raw file](https://github.com/harness/harness-core/blob/04cf7342dd948bad771ae036bb73c2c2e8e85c62/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepOverrideHelper.java#L789)):*
> ```Java
>     }
> 
>     saveFinalConnectionStringsToSweepingOutput(ambiance, connectionStringsSweepingOutputName, finalConnectionStrings);
> ```

The same as above.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java` line 164 at r5](https://reviewable.io/reviews/harness/harness-core/53292#-Nfp7okZ1XcYeDQBcQa6:-Nfp7ok_3sHmAkblB8F3:b5gotjl) ([raw file](https://github.com/harness/harness-core/blob/04cf7342dd948bad771ae036bb73c2c2e8e85c62/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java#L164)):*
> ```Java
>         logStreamingStepClientFactory, prepareServiceAmbiance(ambiance), commandUnit, shouldOpenStream);
>   }
>   private Ambiance prepareServiceAmbiance(Ambiance ambiance) {
> ```

Minor: rename method prepareServiceAmbiance >> prepareServiceorCustomeStageEnvAmbiance
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java` line 175 at r5](https://reviewable.io/reviews/harness/harness-core/53292#-Nfp86zq13EqfmXlMpDN:-Nfp86zq13EqfmXlMpDO:b-2pmq5s) ([raw file](https://github.com/harness/harness-core/blob/04cf7342dd948bad771ae036bb73c2c2e8e85c62/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java#L175)):*
> ```Java
>       }
>     }
>     throw new UnsupportedOperationException("Not inside service step or one of it's children");
> ```

Minor: We should have more general or two error msg one for service and one for custom stage env.

**Comment by lovish1999** [comment]
trigger uts

**Comment by lovish1999** [comment]
trigger uts

**Comment by lovish1999** [comment]
trigger ut6

**Comment by lovish1999** [comment]
trigger ut9

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/CustomStageEnvironmentStep.java` line 162 at r6](https://reviewable.io/reviews/harness/harness-core/53292#-NftKzXt6gRI_99RTWFc:-NftKzXt6gRI_99RTWFd:b-wvyk1d) ([raw file](https://github.com/harness/harness-core/blob/c736339f0fdc89a35851c1708bb340ecefe7bcb9/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/CustomStageEnvironmentStep.java#L162)):*
> ```Java
>         serviceStepOverrideHelper.prepareAndSaveFinalConnectionStringsMetadataToSweepingOutput(
>             ngEnvironmentConfig, ambiance, ServiceStepV3Constants.SERVICE_CONNECTION_STRINGS_SWEEPING_OUTPUT);
>       }
> ```

we should move this to a new method for readability

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/constants/BUILD.bazel` line 13 at r6](https://reviewable.io/reviews/harness/harness-core/53292#-NftL5CKCteRHa4fkiWN:-NftL5CKCteRHa4fkiWO:bpof18w) ([raw file](https://github.com/harness/harness-core/blob/c736339f0fdc89a35851c1708bb340ecefe7bcb9/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/constants/BUILD.bazel#L13)):*
> ```Starlark
>     srcs = glob(["*.java"]),
>     visibility = ["//visibility:public"],
>     deps = ["//127-cd-nextgen-entities:module"],
> ```

do we need to depend on entire module ?

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/helper/PlanCreatorHelper.java` line 28 at r6](https://reviewable.io/reviews/harness/harness-core/53292#-NftLQlX1OqqQS1MrXoh:-NftLQlX1OqqQS1MrXoi:brj1p12) ([raw file](https://github.com/harness/harness-core/blob/c736339f0fdc89a35851c1708bb340ecefe7bcb9/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/helper/PlanCreatorHelper.java#L28)):*
> ```Java
> 
> @Slf4j
> public class PlanCreatorHelper {
> ```

We might want to rename this to be more specific like ServicePlanCreatorHelper
Actually, we already have ServiceAllinOnePlanCreatorUtils, can we re-use that ?

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java` line 210 at r6](https://reviewable.io/reviews/harness/harness-core/53292#-NftLh6B1MsNA-zoTTRn:-NftLh6B1MsNA-zoTTRo:bbmxkcp) ([raw file](https://github.com/harness/harness-core/blob/c736339f0fdc89a35851c1708bb340ecefe7bcb9/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/CustomStagePlanCreator.java#L210)):*
> ```Java
> 
>     // Add manifests node
>     planCreatorHelper.addManifestsNode(planCreationResponseMap, childrenNodeIds);
> ```

we should not modify arguments. I would say we should do this instead

PlanNode node = // method to create manifest node
planCreationResponseMap.put(..)
childrenNodeIds.add(..)

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 2 of 9 files at r1, 2 of 9 files at r2, 1 of 1 files at r4, 4 of 7 files at r5, 1 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53292)* status: 7 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained.

**Review by ivan-83 — COMMENTED** [review]
> Reviewed 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53292)* status: 6 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained.

**Comment by lovish1999** [comment]
trigger uts

**Comment by lovish1999** [comment]
trigger uts

**Comment by lovish1999** [comment]
trigger uts

**Review by ivan-83 — APPROVED** [review]
:lgtm:

> Reviewed 8 of 8 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53292#-:-Nfu_ZsfEuIgNecgM6iZ:bnfp4nl)* status: 10 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java` line 168 at r8](https://reviewable.io/reviews/harness/harness-core/53292#-Nfu_5Vh07wVKqhUcwkM:-Nfu_5Vh07wVKqhUcwkN:bc0273l) ([raw file](https://github.com/harness/harness-core/blob/a27c35ec25350197379d9a11f873c0ba023cba58/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java#L168)):*
> ```Java
>   private Ambiance prepareServiceOrCustomStageEnvAmbiance(Ambiance ambiance) {
>     List<Level> levels = ambiance.getLevelsList();
>     StepType customStageEnvStep = StepType.newBuilder()
> ```

I guess we can't leverage the step from CustomStageEnvironmentStep.STEP_TYPE.

**Review by ivan-83 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53292#-:-Nfuacpr9kK9tiLm58p-:bnfp4nl)* status: 10 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Comment by lovish1999** [comment]
trigger ut2, ut4, ut5, ut7

**Comment by lovish1999** [comment]
trigger ut7

**Review by lovish1999 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53292)* status: 10 file(s) to review. 1 change requests. 5. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/CustomStageEnvironmentStep.java` line 206 at r3](https://reviewable.io/reviews/harness/harness-core/53292#-NfMDEKM5a_dWlrVXee8:-NfvQz5E83OUHrxeHfnq:b8a12n7) ([raw file](https://github.com/harness/harness-core/blob/dff9c602c1f081004ed268632d4cfb0c01f4cbd1/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/CustomStageEnvironmentStep.java#L206)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

Ok, how are we fetching app settings and con strings from git?
</blockquote></details>

As discussed, we will pick this up separately in azure settings step. 
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepOverrideHelper.java` line 789 at r5](https://reviewable.io/reviews/harness/harness-core/53292#-NfpAKnU46s3q1HKvSUp:-NfvR3_9AsyLNa00kzQW:b32w1t) ([raw file](https://github.com/harness/harness-core/blob/04cf7342dd948bad771ae036bb73c2c2e8e85c62/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepOverrideHelper.java#L789)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

The same as above.
</blockquote></details>

Ok
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java` line 168 at r8](https://reviewable.io/reviews/harness/harness-core/53292#-Nfu_5Vh07wVKqhUcwkM:-NfvR5Za6j1QTWbrJ6pr:b-9a5zkb) ([raw file](https://github.com/harness/harness-core/blob/a27c35ec25350197379d9a11f873c0ba023cba58/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java#L168)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

I guess we can't leverage the step from CustomStageEnvironmentStep.STEP_TYPE.
</blockquote></details>

Yes, because of build cycle issues. Because custom stage env. step injects ServiceStepsHelper.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/constants/BUILD.bazel` line 13 at r6](https://reviewable.io/reviews/harness/harness-core/53292#-NftL5CKCteRHa4fkiWN:-NfvRJNRB_zzKhuU__fr:bct65fo) ([raw file](https://github.com/harness/harness-core/blob/c736339f0fdc89a35851c1708bb340ecefe7bcb9/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/steps/constants/BUILD.bazel#L13)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

do we need to depend on entire module ?
</blockquote></details>

Yes, because ExecutionNodeType had its build file as the build file of the whole module. But for now, we have figured out another way and removed the custom stage env. step constants class. 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/helper/PlanCreatorHelper.java` line 28 at r6](https://reviewable.io/reviews/harness/harness-core/53292#-NftLQlX1OqqQS1MrXoh:-NfvRUeSBh0f-fpKkvo5:bt7g62n) ([raw file](https://github.com/harness/harness-core/blob/c736339f0fdc89a35851c1708bb340ecefe7bcb9/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/helper/PlanCreatorHelper.java#L28)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

We might want to rename this to be more specific like ServicePlanCreatorHelper
Actually, we already have ServiceAllinOnePlanCreatorUtils, can we re-use that ?
</blockquote></details>

Okay. Done.

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java` line 158 at r8](https://reviewable.io/reviews/harness/harness-core/53292#-NfyWoNw03QhI4KsnQtr:-NfyWoNw03QhI4KsnQts:b5zaoqm) ([raw file](https://github.com/harness/harness-core/blob/a27c35ec25350197379d9a11f873c0ba023cba58/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java#L158)):*
> ```Java
>   }
> 
>   public NGLogCallback getServiceLogCallback(Ambiance ambiance, boolean shouldOpenStream) {
> ```

why can't we have different methods for service and environment step ?

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 4 files at r6, 8 of 8 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53292)* status: 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by lovish1999 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53292)* status: 1 change requests. 4. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java` line 158 at r8](https://reviewable.io/reviews/harness/harness-core/53292#-NfyWoNw03QhI4KsnQtr:-NfyZa3lE6rebyyr1zrk:brmyoo0) ([raw file](https://github.com/harness/harness-core/blob/a27c35ec25350197379d9a11f873c0ba023cba58/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceStepsHelper.java#L158)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

why can't we have different methods for service and environment step ?
</blockquote></details>

The methods are used by manifest and config files step also where they throw error if executed by custom stage env. step. However we will handle logging in a separate PR next in line.

**Review by yogesh-chauhan — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53292)* status: 1 change requests. 3. unresolved discussions.

**Review by yogesh-chauhan — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53292)* status: 1 change requests. 3. unresolved discussions.

**Comment by lovish1999** [comment]
trigger revierw

**Comment by lovish1999** [comment]
trigger review

**Pull request merged**